### PR TITLE
refactor: use `Arc` instead of making copies of `Flags` struct

### LINKED ISSUE #13311: `deno::Flags` should be reference counted
The [`Flags` struct](https://github.com/denoland/deno/blob/1fb5858009f598ce3f917f9f49c466db81f4d9b0/cli/flags.rs#L211-L256) has a lot of data in it and it is cloned multiple times. It makes sense to wrap it in an `Rc`.

For example:

https://github.com/denoland/deno/blob/1fb5858009f598ce3f917f9f49c466db81f4d9b0/cli/main.rs#L378

https://github.com/denoland/deno/blob/1fb5858009f598ce3f917f9f49c466db81f4d9b0/cli/main.rs#L572

### PR DESCRIPTION
**Pull request by biryukovmaxim**
ProcState.Flags: Flags -> Arc<Flags>
using Arc instead of cloning

closes #13311

### CODE REVIEW

**Comment by crowlKats** [comment]
I'd rather be for calling it `FlagsInner` instead of just `Inner`

File: cli/flags.rs [hunk 1]
```diff
@@ -208,7 +210,23 @@ impl Default for CheckFlag {
 }
 
 #[derive(Clone, Debug, PartialEq, Default)]
-pub struct Flags {
+pub struct Flags(pub(crate) Arc<Inner>);
```
Comment (dsherret) [inline cli/flags.rs:213]: I think that we don't need to change this interface? Probably we only need to wrap it in an `Arc` in a couple places downstream from this file and maybe put it in an `Arc` before it goes into the proc state. Also, it seems like we're just needlessly cloning `Flags` in a few places because we could just use `ps.flags` instead.

**Comment by biryukovmaxim** [comment]
> I'd rather be for calling it `FlagsInner` instead of just `Inner`
I took an example from this implementation of Wrapper
https://github.com/denoland/deno/blob/a7850d7fe6c52225acd614167495d765ed2385f3/cli/proc_state.rs#L63
But actually I can rename Inner to FlagsInner

File: cli/flags.rs [hunk 2]
```diff
@@ -389,13 +407,13 @@ impl Flags {
 impl From<Flags> for PermissionsOptions {
```
Comment (dsherret) [inline cli/flags.rs:407]: Instead of having this as an `impl From<Flags>` we could instead introduce a `permissions_options()` method on `Flags` that returns this information from a reference of `Flags`.

Comment (biryukovmaxim) [reply cli/flags.rs:407]: Yep, I think its a good idea

File: cli/flags.rs (hunk 1 above)
Comment (bartlomieju) [reply cli/flags.rs:213]: I agree, no changes should be necessary to `Flags` struct itself

File: cli/flags.rs [hunk 3]
```diff
@@ -2524,10 +2542,10 @@ mod tests {
   #[test]
   fn upgrade() {
     let r = flags_from_vec(svec!["deno", "upgrade", "--dry-run", "--force"]);
-    let flags = r.unwrap();
+    let inner = Arc::try_unwrap(r.unwrap().0).unwrap();
     assert_eq!(
-      flags,
-      Flags {
+      inner,
```
Comment (ry) [inline cli/flags.rs:2547]: can you not do just `*r.unwrap()` ?

File: core/module_specifier.rs [hunk 4]
```diff
@@ -123,7 +123,7 @@ pub fn resolve_url(
 /// file path; if it is a relative path it's resolved relative to the current
 /// working directory.
 pub fn resolve_url_or_path(
-  specifier: &str,
+    specifier: &str,
```
Comment (ry) [inline core/module_specifier.rs:126]: unrelated change...

File: cli/tools/test.rs [hunk 5]
```diff
@@ -1,6 +1,6 @@
 // Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
 
-use crate::cache;
+use crate::{cache, Flags};
```
Comment (ry) [inline cli/tools/test.rs:3]: We prefer separate lines for each `use`.

File: cli/main.rs [hunk 6]
```diff
@@ -38,7 +38,7 @@ mod windows_util;
 
 use crate::file_fetcher::File;
 use crate::file_watcher::ResolutionResult;
-use crate::flags::BundleFlags;
+use crate::flags::{BundleFlags, Flags};
```
Comment (ry) [inline cli/main.rs:41]: in Deno we prefer separate lines for each use statement.

**Review by ry — COMMENTED** [review]
Just some small nit-picky comments - otherwise looks good! Thanks!

File: cli/flags.rs (hunk 1 above)
Comment (dsherret) [reply cli/flags.rs:213]: @ry I believe we'd only need to wrap it in an `Arc` when putting it in `ProcState` and then change some of the other API to not requiring cloning (ex: https://github.com/denoland/deno/pull/13610#discussion_r800233365)? Then if some other place needs that state they could do `let flags = ps.flags.clone();`?

File: cli/flags.rs (hunk 3 above)
Comment (biryukovmaxim) [reply cli/flags.rs:2547]: Moving out  of an `Arc` is imposible, but I can replace it by somethink like that:

``
r.unwrap().0.as_ref().clone();
``

Comment (bartlomieju) [reply cli/flags.rs:2547]: If you are the only holder of the `Arc` then you can use `try_unwrap` to move out of the arc: https://doc.rust-lang.org/std/sync/struct.Arc.html#method.try_unwrap

Comment (biryukovmaxim) [reply cli/flags.rs:2547]: > If you are the only holder of the `Arc` then you can use `try_unwrap` to move out of the arc: https://doc.rust-lang.org/std/sync/struct.Arc.html#method.try_unwrap

and I did it at line 2545:
`let inner = Arc::try_unwrap(r.unwrap().0).unwrap();`

@bartlomieju there something I'm missing here?

Comment (bartlomieju) [reply cli/flags.rs:2547]: Ah didn't notice. Seems okay 👍

File: cli/flags.rs (hunk 1 above)
Comment (ry) [reply cli/flags.rs:213]: Ok sure - but we should remove the `derive(Clone)` from the Flags struct then to prevent people from copying this.

Comment (biryukovmaxim) [reply cli/flags.rs]: @dsherret @bartlomieju @ry 
actually I did it. Arc only in ProcState

**Comment by dsherret** [comment]
@biryukovmaxim there was a format script that needed to be run (https://github.com/denoland/deno/tree/main/tools#formatjs). I ran it and reduced a few more clones so should be good now.

File: cli/main.rs [hunk 7]
```diff
@@ -526,8 +524,9 @@ async fn install_command(
   let mut preload_flags = flags.clone();
   preload_flags.inspect = None;
   preload_flags.inspect_brk = None;
```
Comment (dsherret) [inline cli/main.rs:526]: The `#[derive(Clone)]` on `Flags` is only still necessary because of this clone here. I wonder if there's a nice way that we could remove `#[derive(Clone)]` on `Flags` and still support this scenario.

**Review by dsherret — APPROVED** [review]
LGTM! Thanks @biryukovmaxim!
